pkg/handler/handler.go

```diff
@@ -55,13 +56,39 @@ type OpenAPIService struct {
 
 	lastModified time.Time
 
-	specBytes []byte
-	specPb    []byte
-	specPbGz  []byte
+	jsonCache  cache
+	protoCache cache
+}
+
+type cache struct {
+	BuildCache func() ([]byte, error)
+	once       sync.Once
+	bytes      []byte
+	etag       string
+	err        error
+}
 
-	specBytesETag string
-	specPbETag    string
-	specPbGzETag  string
+func (c *cache) Get() ([]byte, string, error) {
+	c.once.Do(func() {
+		bytes, err := c.BuildCache()
```

Comment:
I think this will panic if a query comes before the first `UpdateSpec`, given that `BuildCache` starts with a nil pointer

Reply:
Good point - what do you think should be the correct behavior here? If we haven't called `UpdateSpec` yet we don't have a spec to marshal, if I'm following the logic here correctly. It seems like returning a 503 might be the way to go here.

Reply:
I agree that 503 is probably the way to go. Hard to do from here. Maybe always handle errors with 503 (when bytes is nil) and make sure you return an error if `BuildCache` is not set-up yet, WDYT?

Reply:
To avoid the panic, I think we can initialize the caches with a `return nil, errors.New("...")` `BuildCache` function in `NewOpenAPIService`. Yeah resolving the error to a 503 sounds good to me

Reply:
Ah! That's a pretty clean way to do that!

Reply:
Now that I take a second look at the codepaths, is there actually a way to create `OpenAPIService` without passing a spec? `NewOpenAPIService` requires a spec to be passed in, and at that point we can pass the correct functions to `BuildCache`. Using `return nil, errors.New("...")` sounds like a very clean way to solve the issue if there is such a way, though!

(NewOpenAPIService for reference:)
```
// NewOpenAPIService builds an OpenAPIService starting with the given spec.
func NewOpenAPIService(spec *spec.Swagger) (*OpenAPIService, error) {
	o := &OpenAPIService{}
	if err := o.UpdateSpec(spec); err != nil {
		return nil, err
	}
	return o, nil
}
```

Reply:
Oh you're right! I think I was imagining a bootstrap scenario that never existed... Sorry for the confusion!